### PR TITLE
Rack 3 support

### LINKED ISSUE #1797: Support `rack` 3
See https://github.com/rack/rack/blob/main/CHANGELOG.md

Example test run with rack main branch: https://github.com/sinatra/sinatra/runs/7418697522

### PR DESCRIPTION
**Pull request by dentarg**
Adds support for Rack 3, drops support for Rack 2.

Close #1797

- [x] make sinatra test `handles params without a value` pass
- [x] address various rack deprecations
- [x] address `Rack::Handler is deprecated and replaced by Rackup::Handler`
- [x] make [rack `latest` job work](https://github.com/sinatra/sinatra/actions/runs/3808067712/jobs/6478335024#step:7:10)
- [x] restore `continue-on-error` in test workflow
- [x] use rackup 2.x https://rubygems.org/gems/rackup
- [x] use rack-session 2.x https://rubygems.org/gems/rack-session

See the individual commits for various details.

<Details><Summary>Old comment from Dec, 2022</Summary>

There's one failing test left to address (besides async tests that I have disabled for now, that area is being discussed a bit in #1853): EDIT: behaviour changed in Rack 3: https://github.com/rack/rack/pull/1989

https://github.com/sinatra/sinatra/blob/9b6ba81089244826395577113ac7bac622fdc327/test/routing_test.rb#L314-L323

Opening as draft until we have addressed that – and also until I've taken this for a spin with a real app – something for next year! Happy New Year Sinatra users 🥳 

</Details>

---

This work was sponsored by [84codes](https://84.codes/) ([CloudAMQP](https://www.cloudamqp.com/), [CloudKarafka](https://www.cloudkarafka.com/), [ElephantSQL](https://www.cloudmqtt.com/), [CloudMQTT](https://www.elephantsql.com/))

### CODE REVIEW

**Comment by dentarg** [comment]
> There's one failing test left to address

Some links about it:

* https://github.com/sinatra/sinatra/commit/d8c18394d86626726c256ce57c3336ca649a33ae
* https://github.com/sinatra/sinatra/pull/1517
* https://github.com/sinatra/sinatra/pull/1526

Re: async/streaming, https://github.com/dentarg/sinatra/actions/runs/3807969308/jobs/6478153396 is a test run before I reverted https://github.com/sinatra/sinatra/pull/1857/commits/421018419d64ce900052848ea464d778a7b548d9 if anyone are curious

File: .github/workflows/test.yml
Comment (dentarg) [inline .github/workflows/test.yml:60]: This was to get feedback on all three test suites, should be restored before merge (same for the `continue-on-error` for sinatra-contrib tests)

**Comment by jkowens** [comment]
Thank you @dentarg and 84codes 🤗

**Comment by ioquatix** [comment]
It would be great to have a 3.x release that supports Rack 3+

**Comment by dentarg** [comment]
@ioquatix you mean a release that supports both Rack 2 and Rack 3?

**Comment by ioquatix** [comment]
Yeah,

You can use `rack-session` and `rackup` gems if those are dependencies, the v1 release of those gems is for Rack v2 and the v2 release of those gems is for Rack 3+.

**Comment by dentarg** [comment]
We'll see. Sounds like (a lot?) more effort.

**Comment by ioquatix** [comment]
Maybe check this for ideas about how to be compatible with both Rack 2 and Rack 3. The biggest one is dealing with headers. If you don't want to force people to use lower case headers, you can use `Rack::Headers`.

https://github.com/rails/rails/pull/46594#issuecomment-1397833071

**Comment by ioquatix** [comment]
Looking good.

**Comment by dentarg** [comment]
@ioquatix I wanted to make the tests pass on "rack head"  (main branch) too, but https://github.com/rack/rack/commit/1bd0f1597d8f4a90d47115f3e156a8ce7870c9c8 makes a few tests fail (for [sinatra](https://github.com/sinatra/sinatra/actions/runs/4264817990/jobs/7423388170#step:5:24) and [sinatra-contrib](https://github.com/sinatra/sinatra/actions/runs/4264817990/jobs/7423388170#step:6:237)). Not sure how I should handle that. I suspect that change in Rack will be considered a breaking change? Will it be included in Rack 3.1?

**Comment by ioquatix** [comment]
It will be included in Rack 3.1 - it's not breaking in the sense that it's recommended to use that mime type. The previous mime type is now considered legacy/obsolete, but browsers would handle them both as meaning the same thing. Your test should accept both.

**Comment by dentarg** [comment]
Not sure this is fail related to the changes here, maybe more tweaks are needed? Re: https://github.com/sinatra/sinatra/pull/1887#issuecomment-1441177596

```
1) Failure:
IntegrationAsyncTest#test_with_puma_streams_async_0 [/home/runner/work/sinatra/sinatra/test/integration_async_test.rb:[21](https://github.com/sinatra/sinatra/actions/runs/4336782825/jobs/7572318485#step:5:22)]:
--- expected
+++ actual
@@ -1 +1 @@
-["hi!", "hello"]
+["hi!", "hello", "hi!", "hello"]
```

Locally, it works for me to run the sinatra tests on truffle ruby:

```bash
docker run --rm -it -v $(pwd):/app -w /app ghcr.io/graalvm/truffleruby:debian-22.3.1 bash
apt-get update && apt-get install --yes \
          git \
          pandoc \
          nodejs \
          pkg-config \
          libxml2-dev \
          libxslt-dev \
          libyaml-dev
bundle
bundle exec rake
```

**Comment by dentarg** [comment]
> Not sure [this fail](https://github.com/sinatra/sinatra/actions/runs/4336782825/jobs/7572318485#step:5:21) is related to the changes here

The re-run passed: https://github.com/sinatra/sinatra/actions/runs/4336782825/jobs/7572670654

**Review by zzak — APPROVED** [review]
Nice work!! Thank you so much 🍡

**Comment by zerbinidamata** [comment]
Hey all, is this close to being released?

**Comment by dentarg** [comment]
Depends on what's "close"

I think there's a discussion that needs to happen about dropping Rack 2 support 

Also, I will be busy the coming the weeks

**Comment by dentarg** [comment]
@zerbinidamata have you tested this in a real world application? if not, can you?

**Comment by dentarg** [comment]
Re: my comment from December:

> There's one failing test left to address
> EDIT: behaviour changed in Rack 3: https://github.com/rack/rack/pull/1989
> 
> https://github.com/sinatra/sinatra/blob/9b6ba81089244826395577113ac7bac622fdc327/test/routing_test.rb#L314-L323

Just saw https://github.com/rails/rails/pull/47652 and https://github.com/rack/rack/pull/2052

**Comment by ioquatix** [comment]
We have just discussed this within the Rack maintainers, and myself and Jeremy have decided to revert this behaviour as the change was creating impossible downstream situations. We did our best to try and address the issue, but it just ended up snowballing into levels of complexity we were all unhappy with. So, Rack 3.0.7 will probably revert this behaviour, and the release should be out today if possible.

**Comment by dentarg** [comment]
@ioquatix Wow, okay. Was the initial change in Rack 3 initial or was it an oversight?

I guess Rack 3 adoption isn't great yet so it should be fine...

**Comment by dentarg** [comment]
Maybe Sinatra should wait until Rails is fully compatible with Rack 3 :-) (Maybe even an Rails release?)

**Comment by dentarg** [comment]
@ioquatix Where are you discussing btw, GitHub or is there a Slack for Rack maintainers?

**Comment by dentarg** [comment]
> @ioquatix Where are you discussing btw, GitHub or is there a Slack for Rack maintainers?

Answering myself, as I was really close... #rack in the Bundler Slack (for anyone watching, [invite app](https://slack.bundler.io/) seems broken currently unfortunately)

**Comment by dentarg** [comment]
And I'll note that https://github.com/rack/rack/pull/2059 was just opened

**Comment by ioquatix** [comment]
v3.0.7 was just released.

**Comment by dentarg** [comment]
The meat of discussions about the recent rack changes mentioned above seems to have happened in https://github.com/rack/rack/pull/2038 (for anyone interested)

**Comment by ioquatix** [comment]
Thanks for your continued work on this.

**Comment by lucaskanashiro** [comment]
Is there any plan to merge those changes any time soon?

**Comment by dentarg** [comment]
@lucaskanashiro Yes I have plans to do that (when my time permits). Have you run your app with this PR? How did it go?

**Comment by lucaskanashiro** [comment]
> @lucaskanashiro Yes I have plans to do that (when my time permits). Have you run your app with this PR? How did it go?

I do not have a specific app to test this out, I am interested on this because I'd like to move to rack 3 in Debian and Ubuntu since there are some great improvements that I'd like to ship.

**Comment by dentarg** [comment]
> We have just discussed this within the Rack maintainers, and myself and Jeremy have decided to revert this behaviour as the change was creating impossible downstream situations. We did our best to try and address the issue, but it just ended up snowballing into levels of complexity we were all unhappy with. So, Rack 3.0.7 will probably revert this behaviour, and the release should be out today if possible.

It will be introduced again with https://github.com/rack/rack/pull/2098

**Comment by tibbon** [comment]
How can I help this ship? I can't run a branch in my production app.

**Comment by dentarg** [comment]
@tibbon you could try it out locally perhaps?

File: sinatra.gemspec
Comment (ssjr) [inline sinatra.gemspec:51]: `rack-session` is required only in `rack-protection/lib/rack/protection.rb`. maybe removing this line and add to `rack-protection/rack-protection.gemspec`?
```suggestion
```

**Review by epergo — APPROVED** [review]
LGTM. For what is worth, I've deployed a small application with this branch and is working fine.

**Comment by bvogel** [comment]
I can confirm that our app (350 classes, 16k LOC) will work with this PR. I would just confirm that two suggested changes (marked with 👍) should be incorporated. We would in any case wait for an official release to start using this upgrade.

I expect there to be a mayor version change to stick with semver policies?

**Comment by lucaskanashiro** [comment]
> Yes, there will be a new major version (4) of Sinatra for Rack 3 support. In that release support for Rack 2 will be dropped.

Is there any ETA for that?

**Comment by dentarg** [comment]
No really, but I have a desire to not let this PR get over one year old 😂

File: sinatra.gemspec
Comment (dentarg) [reply sinatra.gemspec:51]: Thanks for catching this, not having the `rack-session` dependency in `rack-protection` actually makes it broken. It isn't discovered in the specs due to this line https://github.com/sinatra/sinatra/blob/d66fa62d05de2d10de9bf8d866c054a425c85885/rack-protection/Gemfile#L14

Comment (dentarg) [reply sinatra.gemspec:51]: I will note though, that solving https://github.com/sinatra/sinatra/issues/1945 means adding `rack-session` as a dependency to Sinatra. I think we would like to address that issue in Sinatra 4.0 too.

Comment (dentarg) [reply sinatra.gemspec]: Tests adjusted in main (dcdebe981074f0a79a0d8384d149c3d2db831a76) and I've moved the dependency to rack-protection now.

**Comment by dentarg** [comment]
Rack 3 drops the `rackup` command, it was moved to the `rackup` gem (https://github.com/rack/rackup,  https://rubygems.org/gems/rackup). This PR makes Sinatra depend on the `rackup` gem. 

Should it? 

`rackup` brings the `webrick` gem as a transient dependency.

`Rack::Handler` (`Rackup::Handler`) is used when starting your application without a `config.ru`:

https://github.com/sinatra/sinatra/blob/11119a8b5c92965a418ab201d5a55877282f2885/lib/sinatra/base.rb#L1587-L1600

The Sinatra README mentions `run with rackup` and also `bundle exec ruby myapp.rb`, so having the dependency would mean less disruption for users utilising this.

**Comment by zzak** [comment]
I think just updating the docs is fine, people updating from Rack 2 -> 3 should be aware of the version differences. Otherwise, when their app goes to boot they should get a pretty loud error that `rackup` is missing.

Since we're also dropping support for Rack 2 here, right? I think this makes sense to bump the MAJOR for sinatra.

**Comment by dentarg** [comment]
Yes, we will drop Rack 2 support when introducing Rack 3 support. Yes, it will happen in a new major version, Sinatra 4.0.

@zzak Currently this PR adds the dependency on `rackup` (and impliclity `webrick`), which means everything will continue to work as it has.

Additional work here is needed if we want to avoid the dependency on `rackup` (and`webrick`) (rescue the load error, check if `Rackup` is defined before trying to use it, if not, raise/print nice message to the user).

**Comment by zzak** [comment]
Yeah, last night I started to write how it would be nice if sinatra doesn't depend on rackup that we make it a soft-dependency and lazily load it if possible, then raise a user-friendly message when it's not installed.

**Comment by dentarg** [comment]
> Yeah, last night I started to write how it would be nice if sinatra doesn't depend on rackup that we make it a soft-dependency and lazily load it if possible, then raise a user-friendly message when it's not installed.

I agree, I will explore making these changes in this branch.

**Comment by dentarg** [comment]
> > Yeah, last night I started to write how it would be nice if sinatra doesn't depend on rackup that we make it a soft-dependency and lazily load it if possible, then raise a user-friendly message when it's not installed.
> 
> I agree, I will explore making these changes in this branch.

With the changes I just pushed, it will look like this without `rackup`:

```bash
$ bundle exec ruby simple_app.rb
Sinatra could not start, the "rackup" gem was not found!

Add it to your bundle with:

    bundle add rackup

or install it with:

    gem install rackup


$ echo $?
1
```

**Comment by ioquatix** [comment]
Wow, nice work!

**Pull request merged**
